integration/etcd_test.go

```diff
@@ -0,0 +1,274 @@
+	putProc, err := gexpect.Spawn(putCmd)
+	if err != nil {
+		return err
+	}
+	putExpect := `{"action":"set","node":{"key":"/testKey","value":"bar","`
+	if _, err := putProc.ExpectRegex(putExpect); err != nil {
+		return err
+	}
+	if err = putProc.Close(); err != nil {
+		return err
+	}
+
+	getCmd := `/bin/sh -c "` + strings.Join(gets, " ") + `"`
+	getProc, err := gexpect.Spawn(getCmd)
+	if err != nil {
+		return err
+	}
+	getExpect := `{"action":"get","node":{"key":"/testKey","value":"bar","`
+	if _, err := getProc.ExpectRegex(getExpect); err != nil {
+		return err
+	}
+	return getProc.Close()
+}
+
+type etcdProcessCluster struct {
+	cfg   *etcdProcessClusterConfig
+	procs []*etcdProcess
+}
+
+type etcdProcess struct {
+	cfg   *etcdProcessConfig
+	proc  *gexpect.ExpectSubprocess
+	donec chan struct{} // closed when Interact() terminates
+}
+
+type etcdProcessConfig struct {
+	args        []string
+	dataDirPath string
+	acurl       url.URL
+}
+
+type etcdProcessClusterConfig struct {
+	clusterSize int
+	isClientTLS bool
+	isPeerTLS   bool
+}
+
+func newEtcdProcessCluster(cfg *etcdProcessClusterConfig) (*etcdProcessCluster, error) {
+	etcdCfgs := cfg.etcdProcessConfigs()
+	epc := &etcdProcessCluster{
+		cfg:   cfg,
+		procs: make([]*etcdProcess, cfg.clusterSize),
+	}
+
+	// launch etcd processes
+	for i := range etcdCfgs {
+		proc, err := newEtcdProcess(etcdCfgs[i])
+		if err != nil {
+			epc.Close()
+			return nil, err
+		}
+		epc.procs[i] = proc
+	}
+
+	// wait for cluster to start
+	readyC := make(chan error, cfg.clusterSize)
+	readyStr := "set the initial cluster version"
+	for i := range etcdCfgs {
+		go func(etcdp *etcdProcess) {
+			_, err := etcdp.proc.ExpectRegex(readyStr)
+			readyC <- err
+			etcdp.proc.ReadUntil('\n') // don't display rest of line
+			etcdp.proc.Interact()
+			close(etcdp.donec)
+		}(epc.procs[i])
+	}
+	for range etcdCfgs {
+		if err := <-readyC; err != nil {
+			epc.Close()
+			return nil, err
+		}
+	}
+	return epc, nil
+}
+
+func newEtcdProcess(cfg *etcdProcessConfig) (*etcdProcess, error) {
+	cmd := `/bin/sh -c "../bin/etcd ` + strings.Join(cfg.args, " ") + ` 2>&1 "`
+	child, err := gexpect.Spawn(cmd)
+	if err != nil {
+		return nil, err
+	}
+	child.Capture()
+	return &etcdProcess{cfg: cfg, proc: child, donec: make(chan struct{})}, nil
+}
+
+func (cfg *etcdProcessClusterConfig) etcdProcessConfigs() []*etcdProcessConfig {
+	clientScheme := "http"
+	if cfg.isClientTLS {
+		clientScheme = "https"
+	}
+	peerScheme := "http"
+	if cfg.isPeerTLS {
+		peerScheme = "https"
+	}
+
+	etcdCfgs := make([]*etcdProcessConfig, cfg.clusterSize)
+	initialCluster := make([]string, cfg.clusterSize)
+	for i := 0; i < cfg.clusterSize; i++ {
+		port := etcdProcessBasePort + 2*i
+		curl := url.URL{Scheme: clientScheme, Host: fmt.Sprintf("localhost:%d", port)}
+		purl := url.URL{Scheme: peerScheme, Host: fmt.Sprintf("localhost:%d", port+1)}
+		name := fmt.Sprintf("testname%d", i)
+		dataDirPath := name + ".etcd"
+		os.RemoveAll(dataDirPath)
```

Comment:
From previous bad experience, I suggest checking error here and panicking if it's non-nil, rather than ignoring.